k2/csrc/aux_labels.cc

```diff
@@ -61,4 +136,93 @@ void MapAuxLabels2(const AuxLabels &labels_in,
   start_pos.push_back(num_labels);
 }
 
+void InvertFst(const Fsa &fsa_in, const AuxLabels &labels_in, Fsa *fsa_out,
+               AuxLabels *aux_labels_out) {
```

Comment:
Obviously for output AuxLabels we'll have to have a different data structure as Array2 and the like do not own their data.  What I am thinking is that these kinds of algorithms could be wrapped as a class.  It would have a constructor which would do the work, a function called size() that returns the number of bytes (or elements?) required, and a function called `output(void *data ? int32_t *data?, size_t data_size /*for checking*/, Fs *fsa_out, AuxLabels *aux_labels_out)`.
At least that's an initial proposal.  LMK your thoughts.
Eventually we'll be using Tensors for memory, so we need to think about this now.

Reply:
what about we wrapper each Array type with a template class, e.g.
```
class Array2Storage: Array2{
   private:
         Ptr* data_storage; // allocate memory here for `data` in `Array2`
}
```
and define `AuxLabel = Array2<int32_t, int32_t> *` which is a pointer, then we can pass `&Array2` or `&Array2Storage`